File: frontend/src/app/core/services/billing.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { environment } from '../../../environments/environment';
import { 
  BillingRequest, 
  BillingResponse, 
  BillingSummaryResponse,
  ApiResponse,
  PageResponse 
} from '../../models';

@Injectable({
  providedIn: 'root'
})
export class BillingService {
  private apiUrl = `${environment.apiUrl}/billings`;

  constructor(private http: HttpClient) {}

  getBillings(page = 0, size = 20, sort = 'billDate,desc'): Observable<BillingSummaryResponse[]> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sort', sort);

    return this.http.get<ApiResponse<PageResponse<BillingSummaryResponse>>>(this.apiUrl, { params })
      .pipe(map(response => response.data.content));
  }

  getAllBillings(): Observable<BillingSummaryResponse[]> {
    return this.http.get<ApiResponse<BillingSummaryResponse[]>>(`${this.apiUrl}/all`)
      .pipe(map(response => response.data));
  }

  getBillingById(id: number): Observable<BillingResponse> {
    return this.http.get<ApiResponse<BillingResponse>>(`${this.apiUrl}/${id}`)
      .pipe(map(response => response.data));
  }

  getBillingByBillNum(billNum: number): Observable<BillingResponse> {
    return this.http.get<ApiResponse<BillingResponse>>(`${this.apiUrl}/bill-num/${billNum}`)
      .pipe(map(response => response.data));
  }

  createBilling(request: BillingRequest): Observable<BillingResponse> {
    return this.http.post<ApiResponse<BillingResponse>>(this.apiUrl, request)
      .pipe(map(response => response.data));
  }

  updateBilling(id: number, request: BillingRequest): Observable<BillingResponse> {
    return this.http.put<ApiResponse<BillingResponse>>(`${this.apiUrl}/${id}`, request)
      .pipe(map(response => response.data));
  }

  deleteBilling(id: number): Observable<void> {
    return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/${id}`)
      .pipe(map(response => response.data));
  }

  recordPayment(id: number, amount: number, paymentMode?: string): Observable<BillingResponse> {
    const request = { amount, paymentMode };
    return this.http.post<ApiResponse<BillingResponse>>(`${this.apiUrl}/${id}/payment`, request)
      .pipe(map(response => response.data));
  }

  getUnbilledTripSheets(): Observable<any[]> {
    return this.http.get<ApiResponse<any[]>>(`${environment.apiUrl}/trip-sheets/unbilled`)
      .pipe(map(response => response.data));
  }

  generateBillPreview(clientId: string): Observable<string> {
    return this.http.get<ApiResponse<string>>(`${this.apiUrl}/preview/${clientId}`)
      .pipe(map(response => response.data));
  }
}

// Made with Bob

File: backend/src/main/java/com/bobcarrental/service/BillingService.java
package com.bobcarrental.service;

import com.bobcarrental.dto.request.BillingRequest;
import com.bobcarrental.dto.response.BillingResponse;
import com.bobcarrental.dto.response.BillingSummaryResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDate;
import java.util.List;

/**
 * Service interface for Billing operations
 * 
 * @author Bob Car Rental System
 * @version 1.0
 */
public interface BillingService {

    /**
     * Create a new billing record
     * 
     * @param request Billing creation request
     * @return Created billing
     */
    BillingResponse createBilling(BillingRequest request);

    /**
     * Update an existing billing record
     * 
     * @param id Billing ID
     * @param request Billing update request
     * @return Updated billing
     */
    BillingResponse updateBilling(Long id, BillingRequest request);

    /**
     * Get billing by ID
     * 
     * @param id Billing ID
     * @return Billing details
     */
    BillingResponse getBillingById(Long id);

    /**
     * Get billing by bill number
     * 
     * @param billNo Bill number
     * @return Billing details
     */
    BillingResponse getBillingByBillNo(String billNo);

    /**
     * Get all billings with pagination
     * 
     * @param pageable Pagination parameters
     * @return Page of billings
     */
    Page<BillingSummaryResponse> getAllBillings(Pageable pageable);

    /**
     * Get billings by trip sheet ID
     * 
     * @param tripSheetId Trip sheet ID
     * @return List of billings
     */
    List<BillingSummaryResponse> getBillingsByTripSheet(Long tripSheetId);

    /**
     * Get billings by client ID
     * 
     * @param clientId Client ID
     * @return List of billings
     */
    List<BillingSummaryResponse> getBillingsByClient(Long clientId);

    /**
     * Get billings by date range
     * 
     * @param startDate Start date
     * @param endDate End date
     * @return List of billings
     */
    List<BillingSummaryResponse> getBillingsByDateRange(LocalDate startDate, LocalDate endDate);

    /**
     * Get unpaid billings
     * 
     * @return List of unpaid billings
     */
    List<BillingSummaryResponse> getUnpaidBillings();

    /**
     * Record a payment for a billing
     * 
     * @param id Billing ID
     * @param amount Payment amount
     * @return Updated billing
     */
    BillingResponse recordPayment(Long id, Double amount);

    /**
     * Delete a billing record
     * 
     * @param id Billing ID
     */
    void deleteBilling(Long id);

    /**
     * Check if bill number exists
     * 
     * @param billNo Bill number
     * @return true if exists, false otherwise
     */
    boolean existsByBillNo(String billNo);

    /**
     * Calculate GST for a billing amount
     * 
     * @param amount Base amount
     * @param gstRate GST rate (e.g., 18 for 18%)
     * @param isInterState true for IGST, false for CGST+SGST
     * @return Array [cgst, sgst, igst]
     */
    /**
     * Generate bill content preview for a client
     * Used to show bill preview when client is selected in the form
     * 
     * @param clientId Client ID
     * @return Generated bill preview content
     */
    String generateBillPreview(String clientId);
    Double[] calculateGst(Double amount, Double gstRate, boolean isInterState);
}

// Made with Bob
